src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp

```diff
@@ -0,0 +1,308 @@
+/// 4. Group conditions with the same set of columns into a single read/compute step
+/// 5. Build DAGs for each step:
+///    - DFS from the condition root node:
+///      - If the node was not computed yet, add it to the DAG and traverse its children
+///      - If the node was already computed by one of the previous steps, add it as output for that step and as input for the current step
+///      - If the node was already computed by the current step just stop traversing
+/// 6. Find all outputs of the original DAG
+/// 7. Find all outputs that were computed in the already built DAGs, mark these nodes as outputs in the steps where they were computed
+/// 8. Add computation of the remaining outputs to the last step with the procedure similar to 4
+bool tryBuildPrewhereSteps(PrewhereInfoPtr prewhere_info, const ExpressionActionsSettings & actions_settings, PrewhereExprInfo & prewhere)
+{
+    if (!prewhere_info || !prewhere_info->prewhere_actions)
+        return true;
+
+    Poco::Logger * log = &Poco::Logger::get("tryBuildPrewhereSteps");
+
+    LOG_TRACE(log, "Original PREWHERE DAG:\n{}", prewhere_info->prewhere_actions->dumpDAG());
+
+    /// 1. List all condition nodes that are combined with AND into PREWHERE condition
+    const auto & condition_root = prewhere_info->prewhere_actions->findInOutputs(prewhere_info->prewhere_column_name);
+    const bool is_conjunction = (condition_root.type == ActionsDAG::ActionType::FUNCTION && condition_root.function_base->getName() == "and");
+    if (!is_conjunction)
+        return false;
+    auto condition_nodes = condition_root.children;
+
+    /// 2. Collect the set of columns that are used in the condition
+    std::unordered_map<const ActionsDAG::Node *, NodeInfo> nodes_info;
+    for (const auto & node : condition_nodes)
+    {
+        fillRequiredColumns(node, nodes_info);
+    }
+
+    /// 3. Sort condition nodes by the number of columns used in them and the overall size of those columns
+    /// TODO: not sorting for now because the conditions are already sorted by Where Optimizer
+
+    /// 4. Group conditions with the same set of columns into a single read/compute step
+    std::vector<std::vector<const ActionsDAG::Node *>> condition_groups;
+    for (const auto & node : condition_nodes)
+    {
+        const auto & node_info = nodes_info[node];
+        if (!condition_groups.empty() && nodes_info[condition_groups.back().back()].required_columns == node_info.required_columns)
+            condition_groups.back().push_back(node);    /// Add to the last group
+        else
+            condition_groups.push_back({node}); /// Start new group
+    }
+
+    /// 5. Build DAGs for each step
+    struct Step
+    {
+        ActionsDAGPtr actions;
+        String column_name;
+    };
+    std::vector<Step> steps;
+
+    OriginalToNewNodeMap node_remap;
+
+    for (const auto & condition_group : condition_groups)
```

Comment:
I haven't read very thoughtfully, but anyway I don't really understand why we cannot use `ActionsDAG::split`?

Reply:
ActionsDAG::split() might work here but it is likely to have worse algorithmic complexity: if we have N condition groups in the DAG, then will call split N-1 times and each time it will cut a DAG for one condition and copy the rest of the DAG for all remaining computations. In my implementation we only traverse nodes in the original DAG.

Reply:
maybe we need another method `cut` or smth like that